File: backend/src/services/prediction.ts
import { TrackedActor } from '../types';
import { distance } from '../utils/geometry';

export function computeTTC(vehicle: TrackedActor, pedestrian: TrackedActor): number | null {
  const relativeVx = vehicle.velocityPxPerSec.x - pedestrian.velocityPxPerSec.x;
  const relativeVy = vehicle.velocityPxPerSec.y - pedestrian.velocityPxPerSec.y;
  const relativeSpeed = Math.hypot(relativeVx, relativeVy);
  if (relativeSpeed < 1) return null;
  const relativeDistance = distance(vehicle.center, pedestrian.center);
  return relativeDistance / relativeSpeed;
}

export function computePET(vehicle: TrackedActor, pedestrian: TrackedActor): number | null {
  const crossingDistance = distance(vehicle.center, pedestrian.center);
  const vehicleSpeed = Math.max(Math.hypot(vehicle.velocityPxPerSec.x, vehicle.velocityPxPerSec.y), 1);
  const pedestrianSpeed = Math.max(Math.hypot(pedestrian.velocityPxPerSec.x, pedestrian.velocityPxPerSec.y), 1);
  const vehicleTime = crossingDistance / vehicleSpeed;
  const pedestrianTime = crossingDistance / pedestrianSpeed;
  return Math.abs(vehicleTime - pedestrianTime);
}

export function predictedConflictWithinSeconds(vehicle: TrackedActor, pedestrian: TrackedActor, thresholdPx = 28): boolean {
  for (const v of vehicle.predictedPath) {
    const p = pedestrian.predictedPath.find((x) => x.t === v.t);
    if (!p) continue;
    if (distance(v, p) <= thresholdPx) return true;
  }
  return false;
}

File: backend_py/app/prediction.py
from __future__ import annotations

from .geometry import distance


def compute_ttc(vehicle: dict, pedestrian: dict) -> float | None:
    rvx = vehicle["velocityPxPerSec"]["x"] - pedestrian["velocityPxPerSec"]["x"]
    rvy = vehicle["velocityPxPerSec"]["y"] - pedestrian["velocityPxPerSec"]["y"]
    relative_speed = (rvx * rvx + rvy * rvy) ** 0.5
    if relative_speed < 1:
        return None
    relative_distance = distance(vehicle["center"], pedestrian["center"])
    return relative_distance / relative_speed


def compute_pet(vehicle: dict, pedestrian: dict) -> float | None:
    crossing_distance = distance(vehicle["center"], pedestrian["center"])
    vehicle_speed = max(
        (vehicle["velocityPxPerSec"]["x"] ** 2 + vehicle["velocityPxPerSec"]["y"] ** 2) ** 0.5,
        1,
    )
    ped_speed = max(
        (pedestrian["velocityPxPerSec"]["x"] ** 2 + pedestrian["velocityPxPerSec"]["y"] ** 2) ** 0.5,
        1,
    )
    return abs((crossing_distance / vehicle_speed) - (crossing_distance / ped_speed))


def predicted_conflict_within_seconds(vehicle: dict, pedestrian: dict, threshold_px: float = 28) -> bool:
    p_by_t = {int(x["t"]): x for x in pedestrian.get("predictedPath", [])}
    for v in vehicle.get("predictedPath", []):
        key = int(v["t"])
        p = p_by_t.get(key)
        if not p:
            continue
        if distance(v, p) <= threshold_px:
            return True
    return False
